doc/developer/material.md

```diff
@@ -407,23 +407,43 @@ builder.second( rt, isMaterialTransparent );
 
 
 ## Rendering without using Materials {#non-bsdf-rendering}
-The _Radium Material Library_ and related components are mainly designed to manage Materials as a representation of a 
-_Bidirectional Scattering Distribution function (BSDF)_.
+The _Radium Material Library_ and related components are mainly designed to manage Materials as a representation of a _Bidirectional Scattering Distribution function (BSDF)_.
 
-When rendering, it is sometime useful to compute the final color of an object that do not rely on a bsdf but just 
-on a specific color for each geometry fragment.
+When rendering, it is sometime useful to compute the final color of an object that do not rely on a bsdf but just on a specific color for each geometry fragment.
 
-Even if the Ra::Engine::PlainMaterial could be used for this and assuming that such a way to define material is object 
-specific, the following steps are required :
+To define a custom fragment's color computation shader and use it with application provided parameters, the following steps are required :
 
 1. Develop specific vertex and fragment shaders to compute the fragment color
 2. Build a Ra::Engine::ShaderConfiguration that uses these shaders
 3. Build a render technique that use this configuration
-4. Associate the render technique with a geometry in a Ra::Engine::RenderObject
+4. If the shaders have uniform parameters, implement a specific Ra::Engine::ShaderParameterProvider and associate an instance of the parameter provider to the render technique.
+5. Associate the render technique with a geometry in a Ra::Engine::RenderObject
 
-This could result in the following C++ code to configure a RenderObject.
+Here is an example snippet.
 ~~~{.cpp}
-// 1. Develop specific vertex and fragment shaders to compute the fragment color
+// 1. Implement a parameter provider to provide the uniforms for the shader
+class MyParameterProvider : public Ra:Engine::ShaderParameterProvider {
+public:
+  MyParameterProvider() {}
+  ~MyParameterProvider() {}
+  void updateGL() override {
+    // Method called before drawing each frame in Renderer::updateRenderObjectsInternal.
+    // The name of the parameter corresponds to the shader's uniform name.
+    m_renderParameters.addParameter( "aColorUniform", m_colorParameter );
+    m_renderParameters.addParameter( "aScalarUniform", m_scalarParameter );
+  }
+
+  void setOrComputeTheParameterValues() {
+	// client side computation of the parameters, e.g.
```

Comment:
After some experimentation, why do we need to have double storage of m_colorParameter.
Could this method directly set `m_renderParameter.addParameter("aColorUniform", value)` and the updateGL do nothing ...

Reply:
Hum, I guess we could see this as a CPU/GPU transfer ?
I mean, we should be able to use material even without GL.
I didn't check the code, but maybe that's why this is needed.

Reply:
This class is a `ShaderParameterProvider` so let's say it's linked to GPU Shader.
The `ShaderParameterProvider` class stores parameters on the CPU side as uniform value, ready to be transfered.
We can think of a different design, but as is, Materials are already on the GPU side, even if they can give access to material values without a shader, they don't need to store the value twice.

Reply:
It is not mandatory, when you write a class that derive from `ShaderParameterProvider` to store the parameter twice. All the parameters can be stored only once in the `m_renderParameter` object and the UpdateGL method will do nothing (or only some render-state dependent actions). And this can simplify and improve the performance of many rendering operations.
Meanwhile, random access to a named value (a uniform) from `m_renderParameter` is O(log(n)) while accessing the same named value from an object attribute is O(1). As there are not so many parameters for a shader, it will not be so important but no one know what can happen.
This was also designed to allow different representation between CPU and GPU version of a parameter. This is clearly the case for textures.
If you're refactoring RenderParameter to store also the type (vec3, float, ...), definition domain ([-2, 5], RGB subdomain, ...) and semantic (vector, point, color, ...) into the association name<->value while allowing constant time access, all the Engine classes inheriting from `ShaderParameterProvider` might be refactored as well to remove this double storage. In the current Engine implementation, the CPU version is used to load/save the assets or to edit/change a value while ensuring some constraints related to definition domain or relations between parameters (e.g. energy conservation for bsdf)